onmt/translate/BeamSearcher.lua

```diff
@@ -0,0 +1,298 @@
+      rawIndexes = onmt.utils.Cuda.convert(rawIndexes:double())
+      rawIndexes:add(-1)
+      topIndexes = onmt.utils.Cuda.convert(rawIndexes:double():fmod(vocabSize)) + 1 -- (remainingBatchSize, beamSize)
+    end
+    local beamParents = onmt.utils.Cuda.convert(rawIndexes:int()/vocabSize + 1) -- (remainingBatchSize, beamSize)
+    -- use the top k indexes to select the stepOutputs
+    if t == 1 then -- beamReplicate
+      stepOutputs = beamReplicate(stepOutputs, self.beamSize) -- convert to (origBatchSize * beamSize, *)
+    end
+    stepOutputs = beamSelect(stepOutputs, beamParents) -- (remainingBatchSize * beamSize, *)
+
+    -- judge whether end has been reached use topIndexes (batchSize, beamSize)
+    local remaining = {}
+    local newRemainingBatchSize = 0
+    self.origBatchIdToRemainingBatchId[t] = {}
+    local remainingBatchIdToOrigBatchIdTemp = {}
+    for b = 1, remainingBatchSize do
+      local origBatchId = remainingBatchIdToOrigBatchId[b]
+      local done = true
+      for k = 1, nBest do
+        if topIndexes[b][k] ~= self.endSymbol then
+          done = false
+        end
+      end
+      if not done then
+        newRemainingBatchSize = newRemainingBatchSize + 1
+        self.origBatchIdToRemainingBatchId[t][origBatchId] = newRemainingBatchSize
+        remainingBatchIdToOrigBatchIdTemp[newRemainingBatchSize] = origBatchId
+        table.insert(remaining, b)
+      else
+        self.completedTimeStep[origBatchId] = t
+      end
+    end
+    remainingBatchIdToOrigBatchId = remainingBatchIdToOrigBatchIdTemp
+    table.insert(self.beamParentsHistory, beamParents) -- (remainingBatchSize, beamSize)
+    table.insert(self.beamScoresHistory, beamScores:clone()) -- (remainingBatchSize, beamSize)
+    table.insert(self.stepOutputsHistory, recursiveClone(stepOutputs)) -- newRemainingBatchSize
+    if newRemainingBatchSize ~= remainingBatchSize then -- some batches are finished
+      if #remaining ~= 0 then
+        stepOutputs = rcToFlat(selectBatch(flatToRc(stepOutputs, self.beamSize), remaining))
+        topIndexes = selectBatch(topIndexes, remaining)
+        beamScores = selectBatch(beamScores, remaining)
+      else
+        break
+      end
+    end
+    table.insert(self.topIndexesHistory, topIndexes:clone()) -- newRemainingBatchSize
+    t = t + 1
+  end
+end
+
+--[[ Get beam search predictions.
+
+Parameters:
+  * `k` - if set to k, then the k-th predictions will be returned. It must be smaller than or equal to `nBest` in `BeamSearcher:search`. [1]
+
+Returns:
+  * `predictions` - a table. predictions[b][t] stores the prediction in batch `b` and time step `t`.
+  * `scores` - a table. scores[b] stores the prediction score of batch `b`.
+  * `outputs` - a table. outputs[b][j][t] stores the j-th element in `stepOutputs` produced by `stepFunction` in batch `b` and time step `t`. In the case of an empty prediction, i.e. `predictions[b] == {}`, `outputs[b]` will be nil.
+
+]]
+function BeamSearcher:getPredictions(k)
```

Comment:
The reason is that we may call this function multiple times in order to return the k-th hypothesis, while the search only needs to be called once.